Add Memory support

danielbeardsley opened the pull request:
Support retrieving and using the memory information that XDebug records.

Adding "memory" as cost format option was the easiest way to get it done.

danielbeardsley committed:
Preprocessor.php: break up some long lines

Easier to read / maintian this way.

danielbeardsley committed:
PHP 8 fixes (no undefined properties).

danielbeardsley committed:
Support Memory in cachegrind files.

Xdebug has supported memory for a while, but this tool hadn't been
updated to use it.

Here we add a new "costFormat" option of "bytes of memory".
We pass this through to the pre-processor to have it read the memory
cost field instead of the time cost field.

danielbeardsley committed:
Preprocessor.cpp: read memory usage if asked

Just like the PHP pre-processor, we read the memory if costFormat ==
'bytes'.

jarstelfox commented on library/Preprocessor.php:
```diff
@@ -42,10 +42,11 @@ class Webgrind_Preprocessor
-    static function parse($inFile, $outFile)
+    static function parse($inFile, $outFile, $readMemory = false)
     {
+        $costPattern = $readMemory ? "%*d %d" : "%d";
```
Could you explain the reason for this pattern?
The `%*d` is skipping the number correct?
What number is in the second / third positions?

danielbeardsley replied:
Yes, every "cost line" looks like this:
`4 936 160` == `{line number} {time cost in usec} {memory cost in bytes}`

When we want to read memory, we effectively need to skip the time cost and read the last number (which is normally unread).

andyg0808 reviewed (approved):
CR :+1: 
Looks reasonable.

jarstelfox commented on library/FileHandler.php:
```diff
@@ -160,12 +160,14 @@ public function getTraceList() {
     public function getTraceReader($file, $costFormat) {
-        $prepFile = Webgrind_Config::storageDir().$file.Webgrind_Config::$preprocessedSuffix;
+        $readMemory = $costFormat === 'bytes';
+        $memorySuffix = $readMemory ? ".memory" : "";
+        $prepFile = Webgrind_Config::storageDir().$file.$memorySuffix.Webgrind_Config::$preprocessedSuffix;
         try {
             $r = new Webgrind_Reader($prepFile, $costFormat);
         } catch (Exception $e) {
             // Preprocessed file does not exist or other error
-            Webgrind_Preprocessor::parse(Webgrind_Config::xdebugOutputDir().$file, $prepFile);
+            Webgrind_Preprocessor::parse(Webgrind_Config::xdebugOutputDir().$file, $prepFile, $readMemory);
```
Ah! Cool, you are creating two pre-processed files (one for memory, one for CPU)

jarstelfox reviewed (approved):
CR 🌵

kthaler commented:
QA 💎 - Webgrind now receives and presents memory information.
deploy_block 🦖 - I'm noticing some serious discrepancies between kcachegrind (qcachegrind for me on M1) and Webgrind. For example:

![Screenshot 2023-11-03 at 1 53 22 PM](https://github.com/iFixit/webgrind/assets/145375039/3c0016b9-3069-4c42-9681-459912c57d99)

If this is alright, then please disregard the deploy block.

danielbeardsley commented:
I made a minimal example and the data is coming out correct in both tools for me. Actually, it appears that `kcachegrind` has a bug.

```php
<?

function main() {
   $x[] = useMemory(100);
   $x[] = useMemory(1000);
   $x[] = useMemory(10000);
   $x[] = useMemory(100000);
   $x[] = useMemory(1000000);
   $x[] = useMemory(10000000);
}

function useMemory(int $bytes) {
   return str_repeat("x", $bytes);
}

main();
```

Note that `kcachegrind` says that `main()` has `0` bytes of inclusive cost. This is obviously false, so I'm not sure what's going on. Maybe it's cause I named in `main` like the app entrypoint? I'll try a more complex example.

![image](https://github.com/iFixit/webgrind/assets/26855/b6769335-5a45-4205-85fc-57ae98aa69ee)

danielbeardsley commented:
Nah, it seems like it's a bug in `kcachegrind`

Here's a slightly more complex example:

```php
function main() {
   $x[] = use10KB();
   $x[] = use1MB();
}

function use1MB() {
   $overhead = useMemory(100);
   return useMemory(10000000);
}

function use10KB() {
   $overhead = useMemory(100);
   return useMemory(10000);
}

function useMemory(int $bytes) {
   return str_repeat("x", $bytes);
}

main();
```

![image](https://github.com/iFixit/webgrind/assets/26855/3da0e52d-9e7e-4eac-bcab-53d6e60d6a81)

danielbeardsley commented:
@jarstelfox recommended I change "main" to something else. No dice, `kcachegrind` still shows the bug:

![image](https://github.com/iFixit/webgrind/assets/26855/7551268b-ada3-4450-b7d9-5705b6ddfd52)

danielbeardsley commented:
dev_block :+1: the binary pre-processor is producing the wrong results.

jarstelfox commented:
I do wish there was some test suite here.
You should open this up on the upstream 💙

~~Welp, your examples give me confidence that this is working as expected.~~

looks like you found a bug

danielbeardsley committed:
blockUI: handle jQuery upgrade

Some other pull upgraded jQuery and that caused this jQuery plugin to
crash. Stubbing this "browser" object will allow expressions like
"$.browser.ie" to return undefined instead of throwing.

danielbeardsley committed:
Table sorting: be explicit with column types

Previously, the tablesorter was detecting numeric values with thousands
separators as "text" and sorting them incorrectly.

I added thousands separators for bytes values, so they are easier to
read at a glance as the numbers get long but the sorter hadn't been
parsing them well.

Note: I'm not sure why the "Cost" column was set to "sorter: false" on
the subtables, it had still been sorting but it had been falling back to
auto-detecting the sort type.

danielbeardsley commented:
un_dev_block :+1: The outputs are the same, it was a sorting type auto-detection issue.

jarstelfox reviewed (approved):
CR 🌵

andyg0808 reviewed (approved):
CR :+1:

kthaler commented:
QA 💎  - The bytes of memory cost option in webgrind is displaying correct values.

un_deploy_block 🦕  - Fixed the problem with my qcachegrind and it is now showing the same/similar values as webgrind.
